src/GitHub.Exports.Reactive/Collections/TrackingCollection.cs

```diff
@@ -0,0 +1,889 @@
+                .Select(FilteredNone)
+                .Select(FilteredInsert)
+                .Select(FilteredMove)
+                .TimeInterval()
+                .Select(UpdateProcessingDelay)
+                .Select(data => data.Item)
+                .Publish()
+                .RefCount();
+
+            return source;
+        }
+
+        /// <summary>
+        /// Set a new comparer for the existing data. This will cause the
+        /// collection to be resorted and refiltered.
+        /// </summary>
+        /// <param name="theComparer">The comparer method for sorting, or null if not sorting</param>
+        public void SetComparer(Func<T, T, int> theComparer)
+        {
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            SetAndRecalculateSort(theComparer);
+            SetAndRecalculateFilter(filter);
+        }
+
+        /// <summary>
+        /// Set a new filter. This will cause the collection to be filtered
+        /// </summary>
+        /// <param name="theFilter">The new filter, or null to not have any filtering</param>
+        public void SetFilter(Func<T, int, IList<T>, bool> theFilter)
+        {
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            SetAndRecalculateFilter(theFilter);
+        }
+
+        public IDisposable Subscribe()
+        {
+            if (source == null)
+                throw new InvalidOperationException("No source observable has been set. Call Listen or pass an observable to the constructor");
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            disposables.Add(source.Subscribe());
+            return this;
+        }
+
+        public IDisposable Subscribe(Action<T> onNext, Action onCompleted)
+        {
+            if (source == null)
+                throw new InvalidOperationException("No source observable has been set. Call Listen or pass an observable to the constructor");
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            disposables.Add(source.Subscribe(onNext, onCompleted));
+            return this;
+        }
+
+        public void AddItem(T item)
+        {
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            queue.Enqueue(item);
+        }
+
+        public void RemoveItem(T item)
+        {
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+
+            var position = GetIndexUnfiltered(item);
+            if (position < 0)
+                return;
+
+            var data = new ActionData(TheAction.Remove, item, null, position - 1, position, original);
+            data = CheckFilter(data);
+            data = CalculateIndexes(data);
+            data = SortedRemove(data);
+            data = FilteredRemove(data);
```

Comment:
> If it's a pipeline, shouldn't each method return a new instance of ActionData instead of mutating the passed in instance? I think this calling pattern is really confusing.

Yes, if this was a purely functional thing, it would return a new `ActionData` for every call. I don't do that because I don't want to be creating thousands of objects and putting pressure on the GC when I don't have to.

`ActionData` has non-mutable data, which is the data regarding the item that's being processed and where it is in the sorted, unfiltered list (the item, it's position in the list, where it was before, and a reference to the list itself). This data is built on `ProcessItem`. After that, other information is added to `ActionData` so I don't have to keep looking it up, like whether it's visible according to the current filter, it's index in the filtered list, etc. All this (mutable) information is related to the filtered list (the one that is visible via the `Items` property of the base class).

I tried to strike a balance between immutability and the fact that every item processed would have at least 3 "updates" to it's `ActionData` object (or more), and go through 5 methods, and I don't like needlessly throwing pressure at the GC. Whenever data that is essential to the object is calculated, a new `ActionData` is created. Other data merely updates the existing `ActionData` (for instance, if there's no filter, only the data for the unfiltered list is needed)

> Also, SortedRemove and FilteredRemove actually mutate the current TrackingCollection and do not mutate data.

Neither do the other `SortedXXX` and `FilteredXXX` methods that are called via the `Select` calls in the `Listen` method. By the time they're called, the data in `ActionData` is ready for processing by these methods, so they merely use it and pass it through.

In a future where I can figure out how to include Remove actions into the pipeline that's set up in the `Listen` methods, `SortedRemove` and `FilteredRemove` will have to conform to the same pattern as the other `SortedXX` and `FilteredXX` methods. The only reason they're not included in the `Listen` method list of `Select` calls right now is because they would be noops and I don't see the point in having method calls that currently are known to be noops.